### ISSUE TITLE
twisted.internet.tcp.Port.startListening after stopListening rebinds the port but doesn't make it usable again

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2913]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2913|
|Type|defect|
|Created|2007-11-20 15:05:58Z|
|Branch|https://github.com/twisted/twisted/tree/port-restart-2913-2|

Port.doRead has an early exit for the case where ```disconnecting``` is ```True```.  ```stopListening``` sets ```disconnecting``` to ```1``` but nothing else will ever set it back to ```0```, including ```startListening```.  If you call ```startListening``` on a ```Port``` after stopping it, it will never accept another connection, although if a client connects to it, it will become readable and the reactor (select, at least) will spend as much time as it can trying to accept the connection.

<details><summary>Searchable metadata</summary>

```
trac-id__2913 2913
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_port_restart_2913_2 branches/port-restart-2913-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1195571158000000 1195571158000000
changetime__1209977977000000 1209977977000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22483]) Branching to 'port-restart-2913'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22484]) Fix the problem, with a test.

Refs [#2913](https://github.com/twisted/twisted/issues/2913)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

The solution is simple enough for not refactoring the whole of tcp.py, the I guess that's OK. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

I think `test_restartListening` needs to disconnect the server protocol in `check` (oddly named?) to be completely correct.  I have the feeling that on OS X or Windows this will sometimes cause an unclean reactor error.

Fix looks good, although if `factory.doStop` raises an exception... ;)  I suppose I should file a ticket for problems related to that case.  Ignore it unless you really want to fix it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

That was a good place to fix the problem with factory.doStop, so I did it. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @therve|
|-|-|

It would be nice to have an `@ivar` for `deferred` that explains what it's for, now that even more stuff is touching it.

Deleting the 'else' block with the 'raise' in it in `connectionLost` does not cause any tests to fail, so there's probably some missing test coverage there.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [23409]) Process review comments.

Refs [#2913](https://github.com/twisted/twisted/issues/2913)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Thanks, that's done I think. Back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Some minor things:

  * In the docstring of `Port`:
    * In the first paragraph after the introductory sentence, you replaced ''I'' with ''it'' - I'd go with ''this'' (or ''this class'' - ''this'' being used as an adjective here, with ''class'' being the implicit or explicit noun it is modifying) instead.  
    * Similarly, you exchanged ''my'' for ''its'': I'd use ''a'', because ''its'' doesn't have an obvious antecedent here.
    * Not grammar: the old docstring said ''with the incoming connection as an argument'' but this isn't exactly accurate - it's the address that is passed in.
    * The `disconnecting` attribute could probably use some documentation, especially since it changed here. ;)
    * `connected` is also undocumented.
  * in the docstring of `test_exceptInStop`, _raise_ should be _raises_
  * similarly in the docstring of `test_restartAfterExcept`

Less minor:

  * `test_directConnectionLostCall` doesn't have any asserts in it, which always bothers me a bit.  Would it make sense to do something with `connected` or `disconnected`?  That would kind of suck though, since those aren't part of the interface, so it would be a whitebox test and not applicable to all reactors.  I can't think of a blackbox way to trigger that condition at all, though - of course, that sort of indicates that it isn't real; when can connectionLost be triggered if there's not `deferred` attribute?  Ah - if `stopListening` is called before `startListening` succeeds.  Maybe try triggering it that way?
  * winxp32-py2.5-select, fedora64-py2.4-reactors, server-2k8-x86-py2.5-select, and some other slaves fail some tests tests:
```
===============================================================================
[ERROR]: twisted.test.test_tcp.ListeningTestCase.test_restartAfterExcept

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\test\test_tcp.py", line 316, in close
    serverFactory.protocol.transport.loseConnection()
exceptions.AttributeError: 'NoneType' object has no attribute 'transport'
===============================================================================
[ERROR]: twisted.test.test_tcp.ListeningTestCase.test_restartListening

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\test\test_tcp.py", line 267, in close
    serverFactory.protocol.transport.loseConnection()
exceptions.AttributeError: 'NoneType' object has no attribute 'transport'
-------------------------------------------------------------------------------
```
  * the setuptools slave fails completely, maybe just because the branch is older than setuptools support in Twisted.  It'd be good to merge forward so we can test there.

Sorry for all the new stuff. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [23415]) Branching to 'port-restart-2913-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Ok that should be better. A ticket should be opened to remove duplication in iocpreactor.tcp, but I think that's out of scope.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

  * twisted/test/test_tcp.py
    * in `test_restartAfterExcept`, `oldStopFactory` probably isn't necessary.  You can just delete `serverFactory.stopFactory` and it'll go back to using the class attribute.
    * `test_directConnectionLostCall` docstring says "C{deferred} argument" - does it mean "C{deferred} attribute"?
    * `test_exceptInConnectionLostCall` docstring says "exception is getting to the caller" - "exception is passed through to the caller" would be better.

I'm not sure that raising the exception raised by `doStop` if there is no Deferred to send it to is the right thing, but I guess that was the existing behavior, so preserving it for now is alright.

Comments are all minor, merge when you're ready.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [23519]) Process review comments.

Refs [#2913](https://github.com/twisted/twisted/issues/2913)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `closed`|
|-|-|

(In [23520]) Merge port-restart-2913-2

Author: therve
Reviewers: exarkun, glyph
Fixes [#2913](https://github.com/twisted/twisted/issues/2913)

Allow tcp.Port objects to be restarted, and handle exceptions in stopFactory methods.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
